fabric-client-tags-api-v1/src/client/java/net/fabricmc/fabric/impl/tag/client/ClientTagsLoader.java

```diff
@@ -73,26 +72,42 @@ public static Set<Identifier> loadTag(TagKey<?> tagKey) {
 			}
 		}
 
-		HashSet<Identifier> ids = new HashSet<>();
+		HashSet<Identifier> completeIds = new HashSet<>();
+		HashSet<TagKey<?>> completeTags = new HashSet<>();
+		HashSet<Identifier> immediateChildIds = new HashSet<>();
+		HashSet<TagKey<?>> immediateChildTags = new HashSet<>();
 
 		for (TagEntry tagEntry : tags) {
 			tagEntry.resolve(new TagEntry.ValueGetter<>() {
 				@Nullable
 				@Override
 				public Identifier direct(Identifier id) {
+					immediateChildIds.add(id);
 					return id;
 				}
 
 				@Nullable
 				@Override
 				public Collection<Identifier> tag(Identifier id) {
 					TagKey<?> tag = TagKey.of(tagKey.registry(), id);
-					return ClientTags.getOrCreateLocalTag(tag);
+					immediateChildTags.add(tag);
+					LoadedTag localTag = ClientTagsImpl.getOrCreatePartiallySyncedTag(tag);
+					completeTags.addAll(localTag.completeChildTags);
+					completeTags.add(tag);
+					return localTag.completeIds;
 				}
-			}, ids::add);
+			}, completeIds::add);
 		}
 
-		return Collections.unmodifiableSet(ids);
+		// Ensure that the tag does not refer to itself
+		completeTags.remove(tagKey);
+		immediateChildTags.remove(tagKey);
+
+		return new LoadedTag(Collections.unmodifiableSet(completeIds), Collections.unmodifiableSet(completeTags),
+				Collections.unmodifiableSet(immediateChildTags), Collections.unmodifiableSet(immediateChildIds));
+	}
+
+	public record LoadedTag(Set<Identifier> completeIds, Set<TagKey<?>> completeChildTags, Set<TagKey<?>> immediateChildTags, Set<Identifier> immediateChildIds) {
```

Comment:
The `completeChildTags` are now unnecessary again. I think you can remove `immediate` in front of the other two fields. :wink: